drivers/eeprom/eeprom_mb85rc.c

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT fujitsu_mb85rc
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/eeprom.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(EEPROM_MB85RC, CONFIG_EEPROM_LOG_LEVEL);
+
+struct mb85rc_config {
+	struct i2c_dt_spec i2c;
+	struct gpio_dt_spec wp_gpio;
+	size_t size;
+	size_t pagesize;
+	uint8_t addr_width;
+	bool readonly;
+};
+
+struct mb85rc_data {
+	struct k_mutex lock;
+};
+
+static int mb85rc_write_protect_set(const struct device *dev, int value)
+{
+	const struct mb85rc_config *cfg = dev->config;
+
+	if (!cfg->wp_gpio.port) {
+		return 0;
+	}
+
+	return gpio_pin_set_dt(&cfg->wp_gpio, value);
+}
+
+static uint16_t mb85rc_translate_address(const struct device *dev, off_t offset, uint8_t *addr)
+{
+	const struct mb85rc_config *cfg = dev->config;
+
+	off_t page_offset = offset % cfg->pagesize;
+
+	if (cfg->addr_width > 8) {
+		sys_put_be16(page_offset, addr);
+		addr[0] &= BIT_MASK(cfg->addr_width - 8);
+	} else {
+		addr[0] = page_offset & BIT_MASK(cfg->addr_width);
+	}
+
+	return cfg->i2c.addr + (offset >> cfg->addr_width);
+}
+
+static size_t mb85rc_remaining_len_in_page(const struct device *dev, off_t offset, size_t len)
+{
+	const struct mb85rc_config *cfg = dev->config;
+	off_t page_offset = offset % cfg->pagesize;
+	size_t rem = cfg->pagesize - page_offset;
+
+	if (rem > len) {
+		rem = len;
+	}
+
+	return rem;
+}
+
+static int mb85rc_init(const struct device *dev)
+{
+	const struct mb85rc_config *cfg = dev->config;
+	struct mb85rc_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+
```

Comment:
You need to check if the I2C controller device is ready and error out if not.

Reply:
This is now changed